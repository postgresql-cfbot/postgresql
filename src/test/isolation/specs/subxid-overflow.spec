# Subtransaction overflow
#
# This test is designed to cover some code paths which only occur when
# one transaction has overflowed the subtransaction cache.

setup
{
DROP TABLE IF EXISTS subxids;
CREATE TABLE subxids (subx integer);

CREATE OR REPLACE FUNCTION gen_subxids (n integer)
 RETURNS VOID
 LANGUAGE plpgsql
AS $$
BEGIN
  IF n <= 0 THEN
    INSERT INTO subxids VALUES (0);
    RETURN;
  ELSE
    PERFORM gen_subxids(n - 1);
    RETURN;
  END IF;
EXCEPTION /* generates a subxid */
  WHEN raise_exception THEN NULL;
END;
$$;
}

teardown
{
 DROP TABLE subxids;
 DROP FUNCTION gen_subxids(integer);
}

session s1
step subxid	{ BEGIN; SELECT gen_subxids(100); }
step s1c	{ COMMIT; }

session s2
# move xmax forwards
step xmax	{ BEGIN; INSERT INTO subxids VALUES (1); COMMIT;}
# will see step subxid as still running
step s2sel	{ SELECT count(*) FROM subxids WHERE subx = 0; }

# s2sel will see subxid as still running
permutation subxid xmax s2sel s1c
