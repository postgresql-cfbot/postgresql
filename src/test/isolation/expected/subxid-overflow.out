Parsed test spec with 2 sessions

starting permutation: subxid xmax s2sel s1c
step subxid: BEGIN; SELECT gen_subxids(100);
gen_subxids
-----------
           
(1 row)

step xmax: BEGIN; INSERT INTO subxids VALUES (1); COMMIT;
step s2sel: SELECT count(*) FROM subxids WHERE subx = 0;
count
-----
    0
(1 row)

step s1c: COMMIT;
